File: 3160. Find the Number of Distinct Colors Among the Balls/main.py
# total sam

from collections import defaultdict
from typing import List


class Solution:
    def queryResults(self, limit: int, queries: List[List[int]]) -> List[int]:
        ball_to_color = {}
        color_to_balls = defaultdict(int)
        current_colors = 0
        res = []

        for ball, color in queries:
            if ball not in ball_to_color:
                ball_to_color[ball] = color    
                color_to_balls[color] += 1
                if color_to_balls[color] == 1:
                    current_colors += 1
            else:
                old_color = ball_to_color[ball]
                ball_to_color[ball] = color    
                if old_color != color:
                    color_to_balls[color] += 1
                    if color_to_balls[color] == 1:
                        current_colors += 1
                    color_to_balls[old_color] -= 1
                    if color_to_balls[old_color] == 0:
                        current_colors -= 1
            
            res.append(current_colors)

        return res

File: 3160. Find the Number of Distinct Colors Among the Balls/main.go
// total sam

package main

func queryResults(limit int, queries [][]int) []int {
	ballToColor := make(map[int]int)
	colorToBalls := make(map[int]int)
	currentColors := 0
	res := make([]int, len(queries))

	for i, q := range queries {
		ball, color := q[0], q[1]

		_, exists := ballToColor[ball]

		if !exists {
			ballToColor[ball] = color
			colorToBalls[color]++
			if colorToBalls[color] == 1 {
				currentColors++
			}
		} else {
			oldColor := ballToColor[ball]
			ballToColor[ball] = color
			if oldColor != color {
				colorToBalls[color]++
				if colorToBalls[color] == 1 {
					currentColors++
				}
				colorToBalls[oldColor]--
				if colorToBalls[oldColor] == 0 {
					currentColors--
				}
			}
		}
		res[i] = currentColors
	}

	return res
}
